src/main/kotlin/com/wgslfuzz/semanticspreservingtransformations/GenerateArbitraryExpression.kt

```diff
@@ -574,6 +567,208 @@ private fun generateArbitraryInt(
+                // Cannot create a valid identifier from TypeAlias or Struct scope entry
+                scopeEntry !is ScopeEntry.TypeAlias &&
+                scopeEntry !is ScopeEntry.Struct &&
+                // Remove types that have pointers to them in scope
+                // If not remove can in some cases cause the compiler to fail due to aliasing
+                // Only necessary since counting sort uses pointers
+                !scopeEntry.type.subTypes().any { pointerTypesInScope.contains(it::class) } &&
+                // Only allow scope entries that are either the correct type or contain the correct type via indexing
+                scopeEntry.type
+                    .asStoreTypeIfReference()
+                    .subTypes {
+                        when (it) {
+                            // Remove all runtime sized arrays since cannot pull random element out of them
+                            is Type.Array -> it.elementCount != null
+                            // Remove all pointers that cannot be read from
+                            is Type.Pointer -> it.accessMode == AccessMode.READ || it.accessMode == AccessMode.READ_WRITE
+                            else -> true
+                        }
+                    }.contains(type)
+        }
+
+    if (scopeEntries.isEmpty()) return null
+
+    val scopeEntry = fuzzerSettings.randomElement(scopeEntries) as ScopeEntry.TypedDecl
+
+    val scopeEntryIdentifierExpression =
+        Expression.Identifier(
+            name = scopeEntry.declName,
+        )
+
+    return getRandomExpressionOfType(scopeEntryIdentifierExpression, scopeEntry.type, type, fuzzerSettings)
+        ?: throw AssertionError("Could not find element of type: $type in ${scopeEntry.type}")
+}
+
+private fun getRandomExpressionOfType(
+    expression: Expression,
+    expressionType: Type,
+    requiredType: Type,
+    fuzzerSettings: FuzzerSettings,
+): Expression? {
+    if (requiredType == expressionType) {
+        return expression
+    }
+
+    return when (expressionType) {
+        is Type.Array -> {
+            val elementCount =
+                expressionType.elementCount?.let { Expression.IntLiteral(fuzzerSettings.randomInt(it).toString()) }
+                    ?: return null
+
+            getRandomExpressionOfType(
+                expression =
+                    Expression.IndexLookup(
+                        target = expression,
+                        index = elementCount,
+                    ),
+                expressionType = expressionType.elementType,
+                requiredType = requiredType,
+                fuzzerSettings = fuzzerSettings,
+            )
+        }
+        is Type.Vector -> {
+            getRandomExpressionOfType(
+                expression =
+                    Expression.IndexLookup(
+                        target = expression,
+                        index = Expression.IntLiteral(fuzzerSettings.randomInt(expressionType.width).toString()),
```

Comment:
If not too much trouble, it would be good to have the option of indexing with [0], [1], [2], etc., and also with .x, .y, .z. (and also rgba, srgb or whatever other options there are)

Reply:
Or alternatively have this as a separate transformation, as discussed.